WalletWasabi/WabiSabi/WabiSabiCoordinator.cs

```diff
@@ -122,45 +121,50 @@ public void BanDescendant(object? sender, Block block)
 
 	public void BanDoubleSpenders(object? sender, Transaction tx)
 	{
-		bool IsWasabiCoinjoin(uint256 txId)
+		var txId = tx.GetHash();
+		if (IsWasabiCoinJoin(txId, tx))
 		{
-			var finishedCoinJoinIds = Arena.RoundStates
-				.Select(x => x.CoinjoinState)
-				.OfType<SigningState>()
-				.Where(x => x.IsFullySigned)
-				.Select(x => x.CreateUnsignedTransaction().GetHash());
-
-			return CoinJoinIdStore.Contains(txId) || finishedCoinJoinIds.Contains(txId);
+			return;
 		}
 
-		// We don't ban our own coinjoin outputs.
-		if (IsWasabiCoinjoin(tx.GetHash()))
+		var inputOutPoints = tx.Inputs.Select(x => x.PrevOut);
+		var disruptedRounds = Arena.GetRoundsContainingOutpoints(inputOutPoints);
+
+		// No round was disrupted by the received transaction. Nothing to do here.
+		if (disruptedRounds.Length == 0)
 		{
 			return;
 		}
 
-		var outPoints = tx.Inputs.Select(x => x.PrevOut);
-
-		// Detect and punish double spending coins
-		var disrupters = Arena.RoundStates
-			.Where(r => r.Phase != Phase.Ended)
-			.SelectMany(r => r.CoinjoinState.Inputs.Select(a => (RoundId: r.Id, Coin: a)))
-			.Where(x => outPoints.Any(outpoint => outpoint == x.Coin.Outpoint))
-			.ToArray();
-
-		foreach (var (roundId, offender) in disrupters)
+		// Ban all outputs created by the the received transaction because it has spent coins participating in coinjoin rounds.
+		foreach (var indexedOutput in tx.Outputs.AsIndexedOutputs())
 		{
-			Warden.Prison.DoubleSpent(offender.Outpoint, offender.Amount, roundId);
+			Warden.Prison.DoubleSpent(new OutPoint(tx, indexedOutput.N), indexedOutput.TxOut.Value, disruptedRounds);
```

Comment:
Imagine we receive a transaction that destroys 3 inputs and creates 2 outputs. If any of those inputs is participating in a coinjoin round, that round will fail at broadcasting time because the coinjoin transaction will be seen as a double-spending attempt by our bitcoin node.

Given an input can participate in only one round, when that input is spent only one round is disrupted. However, in this imaginary transaction we have 3 inputs, each of which could be participating in three different rounds and could disrupt those three rounds. If one day we see a transaction disrupting more than one round we can be sure it is an attack because such a tx is hard to build with Wasabi currently (attackers don't follow the rules).

Finally, we detect a transaction that disrupts a round. What to do? Banning the inputs make no sense, they are already spent after all and can not be used to attack the coordinator anymore. We have to ban the created outputs, all of them proportionally to their values, I think.

Reply:
It makes sense IMO but if your comment is true `we can be sure it is an attack` it means it would be better to create an offense for this specific case with a high ban time.